### PR TITLE
Implemented MiScript 2

### PR DESCRIPTION
**Pull request by GerardSmit**

### CODE REVIEW

File: src/Miki/Modules/CustomCommands/Services/CustomCommandsService.cs [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+            
+            return true;
+        }
+        
+        /// <summary>
+        /// Create the global context for the runtime.
+        /// </summary>
+        private async ValueTask<ScriptGlobal> CreateGlobalAsync(IContext e, IScriptValue say, bool isDonator)
+        {
+            var keyLimit = isDonator ? KeyLimit : DonatorKeyLimit;
```
Comment (GerardSmit) [inline src/Miki/Modules/CustomCommands/Services/CustomCommandsService.cs]: This should be swapped.

File: src/Miki/Modules/CustomCommands/JsonUtils.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+using System.IO;
+using System.Threading.Tasks;
+using MiScript.Values;
+using MiScript.Values.Literals;
+using MiScript.Values.Numbers;
+using Newtonsoft.Json;
+
+namespace Miki.Modules.CustomCommands
+{
+    public class JsonUtils
```
Comment (GerardSmit) [inline src/Miki/Modules/CustomCommands/JsonUtils.cs]: This class should be moved to the repository MikiStd.

File: src/Miki/Extensions/StringExtensions.cs [hunk 3]
```diff
@@ -11,6 +13,16 @@ public static string AsCode(this object str)
             return $"`{str}`";
         }
 
+        public static string AsCodeBlock(this object str, string language = null)
+        {
+            return $"```{language}\n{str}\n```";
+        }
+
+        public static EmbedBuilder AddCodeBlock(this EmbedBuilder builder, object str, string language = null)
```
Comment (velddev) [inline src/Miki/Extensions/StringExtensions.cs]: ""string extensions""

File: src/Miki/Modules/CustomCommands/CommandHandlers/CustomCommandsHandler.cs [hunk 4]
```diff
@@ -74,48 +23,53 @@ public async ValueTask CheckAsync(IDiscordMessage data, IMutableContext e, Func<
...
-                var db = e.GetService<IUnitOfWork>();
-                var repository = db.GetRepository<CustomCommand>();
+                var success = await service.ExecuteAsync(e, commandName);
 
-                var command = await repository.GetAsync((long)guild.Id, commandName);
-                if (command != null)
+                if (!success)
                 {
-                    tokens = new Tokenizer().Tokenize(command.CommandBody);
+                    await next();
```
Comment (velddev) [inline src/Miki/Modules/CustomCommands/CommandHandlers/CustomCommandsHandler.cs:44]: If not successful shouldn't it throw an exception?

File: src/Miki/Modules/CustomCommands/CommandHandlers/CustomCommandsHandler.cs [hunk 5]
```diff
@@ -74,48 +23,53 @@ public async ValueTask CheckAsync(IDiscordMessage data, IMutableContext e, Func<
...
                 }
             }
+            catch (MiScriptLimitException ex)
+            {
+                var type = ex.Type switch
+                {
+                    LimitType.Instructions => "instructions",
```
Comment (velddev) [inline src/Miki/Modules/CustomCommands/CommandHandlers/CustomCommandsHandler.cs:51]: Could be localized in the future.

File: src/Miki/Modules/CustomCommands/Values/ScriptSayFunction.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace Miki.Modules.CustomCommands.Values
+{
+    public class ScriptSayFunction : ScriptBaseFunction
+    {
+        public ScriptEmbedBuilder EmbedBuilder { get; set; }
+        
+        public StringBuilder Output = new StringBuilder();
+
+        public override Task<IScriptValue> InvokeAsync(Context context, IScriptValue thisValue, IScriptValue[] args, CancellationToken token = default)
```
Comment (velddev) [inline src/Miki/Modules/CustomCommands/Values/ScriptSayFunction.cs]: long line here, maybe put the params on a new line

**Review by velddev — APPROVED** [review]
Check out my comments, looks good

File: src/Miki/Modules/CustomCommands/CommandHandlers/CustomCommandsHandler.cs (hunk 4 above)
Comment (GerardSmit) [reply src/Miki/Modules/CustomCommands/CommandHandlers/CustomCommandsHandler.cs:44]: The return value is if the command was executed, if a custom command doesn't exists it returns false. It throws an exception whenever something went wrong and that's being handled a few lines down.

**Pull request merged**